Editorial pass. (#197)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

with security against pre-computation attacks upon server compromise.
In addition, the protocol provides forward secrecy and the ability to
hide the password from the server, even during password registration.
This document specifies the core OPAQUE protocol and one instantiation
based on 3DH.

--- middle

# Introduction {#intro}

> Password authentication is ubiquitious in many applications. In a common

Reviewer: ```suggestion
Password authentication is ubiquitous in many applications. In a common
```

implementation, a client authenticates to a server by sending its client
ID and password to the server over a secure connection. This makes
the password vulnerable to server mishandling, including accidentally
logging the password or storing it in cleartext in a database. Server
compromise resulting in access to these plaintext passwords is not an
uncommon security incident, even among security-conscious companies.

> Moreover, plaintex password authentication over secure channels like

Author: ```suggestion
Moreover, plaintext password authentication over secure channels like
```

> TLS are also vulnerable to cases where TLS may fail, including: PKI

Reviewer: ```suggestion
TLS is also vulnerable to cases where TLS may fail, including PKI
```

transmitting the salt may require additional protocol messages.

This document describes OPAQUE, a PKI-free secure aPAKE that is secure
against pre-computation attacks. OPAQUE provides forward secrecy (essential
for protecting past communications in case of password leakage) and the
ability to hide the password from the server, even during password
registration. Furthermore, OPAQUE enjoys good performance and an array
of additional features including the ability to increase
the difficulty of offline dictionary attacks via iterated hashing
or other hardening schemes, and offloading these operations to the
client (that also helps against online guessing attacks); extensibility of
the protocol to support storage and retrieval of client secrets solely
based on a password; being amenable to a multi-server distributed
implementation where offline dictionary attacks are not possible without
breaking into a threshold of servers (such a distributed solution requires

> no change or awareness on the client side relative to a single-server implementation).

Reviewer: ```suggestion
no change or awareness on the client-side relative to a single-server implementation).
```

OPAQUE is defined and proven as the composition of two functionalities:
an oblivious pseudorandom function (OPRF) and an authenticated key
exchange (AKE) protocol. It can be seen
as a "compiler" for transforming any suitable AKE protocol into a secure
aPAKE protocol. (See {{security-considerations}} for requirements of the
OPRF and AKE protocols.) This document specifies one OPAQUE instantiation
based on 3DH {{SIGNAL}}. Other instantiations are possible, as discussed in
{{alternate-akes}}, but their details are out of scope for this document.
In general, the modularity of OPAQUE's design makes it easy to integrate
with additional AKE protocols, e.g., TLS, and with future ones such as those
based on post-quantum techniques.

OPAQUE consists of two stages: registration and authenticated key exchange.
In the first stage, a client registers its password with the server and stores
its encrypted credentials on the server. In the second stage, a client obtains
those credentials, recovers them using the client's password, and subsequently
uses them as input to an AKE protocol.

Currently, the most widely deployed PKI-free aPAKE is SRP {{?RFC2945}}, which is

> vulnerable to pre-computation attacks, lacks proof of security, and is less efficient

Reviewer: I'm not sure about the oxford comma here 🤔

- Client (C): Entity that has knowledge of a password and wishes to authenticate.

> - Server (S): Entity that authenticates clients using passwords.

Reviewer: Are we using the C/S notation somewhere?

    This function also needs to satisfy collision resistance.

OPAQUE additionally depends on an Authenticated Key Exchange (AKE) protocol.
This specification defines one particular AKE based on 3DH; see {{ake-protocol}}.
We let `Npk` and `Nsk` denote the size of public and private keys, respectively,
used in the AKE. The AKE protocol must provide the following functions:

- RecoverPublicKey(private_key): Recover the public key related to the input `private_key`.

> - DeriveAuthKeyPair(seed): Derive a private and public authentication key key pair

Reviewer: ```suggestion
- DeriveAuthKeyPair(seed): Derive a private and public authentication key pair
```

  deterministically from the input `seed`.
- GenerateKeyPair(): Return a randomly generated private and public key pair. This can be
  implemented by generating a random private key `sk`, then computing `pk = RecoverPublicKey(sk)`.

Finally, all random nonces used in this protocol are of length `Nn` = 32 bytes.

# Protocol Overview {#protocol-overview}

OPAQUE consists of two stages: registration and authenticated key exchange.
In the first stage, a client registers its password with the server and stores
its encrypted credentials on the server. The client inputs its credentials,

> which includes its password and user identifier, and the server inputs its

Reviewer: > which includes its password and user identifier

We may want to clarify here that this identifier may or may not be the same as used in the protocol

Author: Hmm, maybe. Can you propose text?

> which includes its password and user identifier, and the server inputs its

Reviewer: The phrasing may exclude the external mode. Is it intentional?

Author: No, but I didn't want to confuse reads so early on. I figured omitting that for now and explaining internal/external mode later was easiest.

parameters, which includes its private key and other information. The client
output of this stage is a single value `export_key` that the client may use

> for application-specific purposes, e.g., to encrypt additional information
> to the server. The server output of this stage is a record corresponding to

Reviewer: > to encrypt additional information to the server
this phrase can be misleading and interpreted as "to encrypt additional information sent to the server who can decrypt it".

```suggestion
for application-specific purposes, e.g., to encrypt additional information
to be sent to and saved by server. The server output of this stage is a record corresponding to
```

Author: I went with a slightly different variant: "to encrypt additional information under the password"

identifier, and the server inputs its parameters and password file record
corresponding to the client. The client outputs two values, an `export_key`
(matching that from registration) and a `session_key`, the latter of which
is the primary AKE output. The server outputs a single value `session_key`
that matches that of the client. Upon completion, clients and servers can
use these values as needed.

The authenticated key exchange flow is shown below:

~~~
    creds                             (parameters, record)
      |                                         |
      v                                         v
    Client                                    Server
    ------------------------------------------------
                   AKE message 1
             ------------------------->
                   AKE message 2
             <-------------------------
                   AKE message 3
             ------------------------->
   ------------------------------------------------
      |                                         |
      v                                         v
(export_key, session_key)                  session_key
~~~

> The rest of this document describes the details of these stages in detail.

Reviewer: ```suggestion
The reaminder of this document describes the details of these stages in detail.
```

{{client-credential-storage}} describes how client credential information is

> encoded, encrypted, and stored on the server. {{offline-phase}} describes the

Reviewer: ```suggestion
generated, encoded, encrypted, and stored on the server. {{offline-phase}} describes the
```

to instantiate OPAQUE using different cryptographic dependencies and parameters.

# Client Credential Storage {#client-credential-storage}

OPAQUE makes use of a structure `Envelope` to manage client credentials.
This envelope holds information about its format and content for the client to
obtain its authentication material.

OPAQUE allows applications to either provide custom client private and public keys
for authentication or to generate them internally, making the application oblivious
to the client's private key. Each public and private key value is an opaque byte
string, specific to the AKE protocol in which OPAQUE is instantiated.
These two options are defined as the `internal` and `external` modes.
See {{envelope-modes}} for their specifications.

Applications may pin key material to identities if desired. If no identity is given
for a party, its value MUST default to its public key. The following types of
application credential information are considered:

- client_private_key: The encoded client private key for the AKE protocol.
- client_public_key: The encoded client public key for the AKE protocol.

> - server_public_key: The encoded server public key for the AKE protocol.

Reviewer: If we mention the client's private key in this part, it makes sense to also mention the server's private key. I suggest either both or none (as they are not used as identities).

Reviewer: I mean the private keys

Author: I understand, but since the server private key isn't used for credentials, it's not included.

Output:
- cleartext_credentials, a CleartextCredentials structure

Steps:
1. if server_identity == nil
2.    server_identity = server_public_key
3. if client_identity == nil
4.    client_identity = client_public_key
5. Create CleartextCredentials cleartext_credentials
   with (server_public_key, server_identity, client_identity)
6. Output cleartext_credentials
~~~

During protocol execution, the identity values can be stored in an
implementation-specific `Credentials` object with names matching the
values.

~~~
struct {
  uint8 server_identity;
  uint8 client_identity;
} Credentials;
~~~

## Envelope Structure {#envelope-structure}

A client `Envelope` is constructed based on the `EnvelopeMode`, consisting
of an `InnerEnvelope` entry whose structure is determined by the mode. Future
modes MAY introduce alternate `InnerEnvelope` contents. `Envelope` is
defined as follows:

~~~
struct {
  uint8 nonce[Nn];
  InnerEnvelope inner_env;
  uint8 auth_tag[Nm];
} Envelope;
~~~

nonce: A unique nonce of length `Nn` used to protect this Envelope.

auth_tag: Authentication tag protecting the contents of the envelope, covering the envelope nonce,

> `InnerEnvelope` and `CleartextCredentials`.

Reviewer: ```suggestion
`InnerEnvelope`, and `CleartextCredentials`.
```

(export_key, session_key)                  session_key
~~~

The rest of this document describes the details of these stages in detail.
{{client-credential-storage}} describes how client credential information is
encoded, encrypted, and stored on the server. {{offline-phase}} describes the
first registration stage of the protocol, and {{online-phase}} describes the
second authentication stage of the protocol. {{configurations}} describes how
to instantiate OPAQUE using different cryptographic dependencies and parameters.

# Client Credential Storage {#client-credential-storage}

OPAQUE makes use of a structure `Envelope` to manage client credentials.
This envelope holds information about its format and content for the client to
obtain its authentication material.

OPAQUE allows applications to either provide custom client private and public keys
for authentication or to generate them internally, making the application oblivious
to the client's private key. Each public and private key value is an opaque byte
string, specific to the AKE protocol in which OPAQUE is instantiated.

These two options are defined as the `internal` and `external` modes.

> See {{envelope-modes}} for their specifications.

Reviewer 2: We can probably merge these two paragraphs into one

Applications may pin key material to identities if desired. If no identity is given
for a party, its value MUST default to its public key. The following types of

> application credential information are hereby considered:

Reviewer 2: ```suggestion
application credential information are considered:
```

Steps:
1. auth_key = Expand(random_pwd, concat(envelope.nonce, "AuthKey"), Nh)
2. export_key = Expand(random_pwd, concat(envelope.nonce, "ExportKey", Nh)

> 3. client_private_key, client_public_key =

Reviewer 2: ```suggestion
3. (client_private_key, client_public_key) =
```

## Envelope Modes {#envelope-modes}

The `EnvelopeMode` specifies the structure and encoding of the
corresponding `InnerEnvelope`. This document specifies the values
of the two aforementioned modes:

~~~
enum {
  internal(1),
  external(2),
  (255)
} EnvelopeMode;

> ~~~

Reviewer 2: Now that `EnvelopeMode` is no longer encoded anywhere, it seems like we do not need to define this enum anymore, right? An implementation can choose to use whatever number it desires to represent "internal" and "external", it doesn't have to be 1 and 2, right?

Author: It doesn't, but let's leave this for a separate PR.

Each `EnvelopeMode` defines its own `InnerEnvelope` structure and must implement
the following interface:

- `inner_env, client_public_key = BuildInnerEnvelope(random_pwd, nonce, client_private_key)`:

>   build and return the mode's `InnerEnvelope` structure and the client's public key.

Reviewer 2: ```suggestion
  Build and return the mode's `InnerEnvelope` structure and the client's public key.
```

- `client_private_key, client_public_key = RecoverKeys(random_pwd, nonce, inner_env)`:

>   recover and return the client's private and public keys for the AKE protocol.

Reviewer 2: ```suggestion
  Recover and return the client's private and public keys for the AKE protocol.
```

The implementations of this interface for both `internal` and `external` modes
are in {{internal-mode}} and {{external-mode}}, respectively.

The size of the envelope may vary between modes. If applications implement
{{preventing-client-enumeration}}, they MUST use the same envelope mode throughout
their lifecycle in order to avoid activity leaks due to mode switching.

### Internal mode {#internal-mode}

In this mode, the client's private and public keys are deterministically derived
from the OPRF output.

With the internal key mode the `EnvelopeMode` value MUST be `internal` and the

> `InnerEnvelope` is empty, and the size `Ne` of the serialized `Envelope` is Nn + Nm.

Reviewer 2: ```suggestion
`InnerEnvelope` is empty, and the size `Ne` of the serialized `Envelope` is `Nn + Nm`.
```

To generate the private key OPAQUE-3DH implements `DeriveAuthKeyPair(seed)` as follows:

~~~
DeriveAuthKeyPair(seed)

Parameter:

> - dst, domain separation tag for HashToScalar set to "OPAQUE-HashToScalar".

Reviewer 2: Can we just omit this as a parameter, and hard-code "OPAQUE-HashToScalar" on line 707?

Reviewer 2: So that line 707 reads:

1. private_key = HashToScalar(seed, dst = "OPAQUE-HashToScalar")

3. Output (private_key, public_key)
~~~

HashToScalar(msg, dst) is as specified in {{I-D.irtf-cfrg-voprf}},

> except that dst = "OPAQUE-HashToScalar".

Reviewer 2: Perhaps remove the "except that dst = "OPAQUE-HashToScalar"", since this is already specified above?

~~~
BuildInnerEnvelope(random_pwd, nonce, client_private_key)

Input:
- random_pwd, randomized password.
- nonce, a unique nonce of length `Nn`.
- client_private_key, empty value. Not used in this function,
  it only serves to comply with the API.

Output:
- inner_env, nil value (serves to comply with the API).
- client_public_key, the client's AKE public key.

Steps:
1. seed = Expand(random_pwd, concat(nonce, "PrivateKey"), Nsk)

> 2. _, client_public_key = DeriveAuthKeyPair(seed)

Reviewer 2: ```suggestion
2. (_, client_public_key) = DeriveAuthKeyPair(seed)
```

parameter, as it is not used.

~~~
RecoverKeys(random_pwd, nonce, inner_env)

Input:
- random_pwd, randomized password.
- nonce, a unique nonce of length `Nn`.
- inner_env, an InnerEnvelope structure. Not used in this
  function, it only serves to comply with the API.

Output:
- client_private_key, The encoded client private key for the AKE protocol
- client_public_key, The encoded client public key for the AKE protocol

Steps:
1. seed = Expand(random_pwd, concat(nonce, "PrivateKey"), Nsk)
2. client_private_key, client_public_key = DeriveAuthKeyPair(seed)
4. Output (client_private_key, client_public_key)
~~~

Note that implementations are free to leave out the `inner_env` parameter,
as it is not used.

### External mode {#external-mode}

This mode allows applications to import custom keys for the client. This
specification only imports the client's private key and internally recovers the
corresponding public key. Implementations are free to import both, in which case
the functions `FinalizeRequest()`, `CreateEnvelope()`, and `BuildInnerEnvelope()`
must be adapted accordingly.

With the external key mode the `EnvelopeMode` value MUST be `external`, and the

> size `Ne` of the serialized `Envelope` is Nn + Nm + Nsk.

Reviewer 2: ```suggestion
size `Ne` of the serialized `Envelope` is `Nn + Nm + Nsk`.
```

An encryption key is generated from the hardened OPRF output and used to encrypt
the client's private key, which is then stored encrypted in the `InnerEnvelope`.
This encryption must follow the requirements in {{envelope-encryption}}. On key
recovery, the client's public key is recovered using the private key.

~~~
struct {
  uint8 encrypted_creds[Nsk];
} InnerEnvelope;
~~~

encrypted_creds : Encrypted client_private_key. Authentication of this field is

> ensured with the `AuthTag` in the envelope that covers this `InnerEnvelope`.

Reviewer 2: ```suggestion
ensured with the `auth_tag` field in the envelope that covers this `InnerEnvelope`.
```

If the implementation provides the `client_public_key`, then `BuildInnerEnvelope()`
can skip the `RecoverPublicKey()` call.

~~~
BuildInnerEnvelope(random_pwd, nonce, client_private_key)

Input:
- random_pwd, randomized password.
- nonce, a unique nonce of length `Nn`.

> - client_private_key, The encoded client private key for the AKE protocol.

Reviewer 2: ```suggestion
- client_private_key, the encoded client private key for the AKE protocol.
```

RecoverKeys(random_pwd, nonce, inner_env)

Input:
- random_pwd, randomized password.
- nonce, a unique nonce of length `Nn`.
- inner_env, an InnerEnvelope structure.

Output:

> - client_private_key, The encoded client private key for the AKE protocol.

Reviewer 2: ```suggestion
- client_private_key, the encoded client private key for the AKE protocol.
```

# Offline Registration {#offline-phase}

This section describes the registration flow, message encoding, and helper functions.
In a setup phase, the client chooses its password, and the server chooses its own pair
of private-public AKE keys (server_private_key, server_public_key) for use with the
AKE, along with a Nh-byte oprf_seed. The server can use the same pair of keys with multiple
clients and can opt to use multiple seeds (so long as they are kept consistent for
each client). These steps can happen offline, i.e., before the registration phase.

> If using `external` mode, the client provides a key pair

Reviewer: ```suggestion
If using the `external` mode, the client provides a key pair
```

The server inputs the following values:

- server_private_key: server private key for the AKE protocol.
- server_public_key: server public key for the AKE protocol.
- credential_identifier: client credential identifier.
- oprf_seed: seed used to derive per-client OPRF keys.

The registration protocol then runs as shown below:

~~~
  Client                                         Server
 ------------------------------------------------------
 (request, blind) = CreateRegistrationRequest(password)

                        request
              ------------------------->

(response, oprf_key) = CreateRegistrationResponse(request,
                          server_public_key,
                          credential_identifier,
                          oprf_seed)

                        response
              <-------------------------

 (record, export_key) = FinalizeRequest(client_private_key,
                                        password,
                                        blind,
                                        response)

                        record
              ------------------------->
~~~

{{registration-functions}} describes details of the functions and the

> correspoding parameters referenced above.

Reviewer: ```suggestion
corresponding parameters referenced above.
```

In a setup phase, the client chooses its password, and the server chooses its own pair
of private-public AKE keys (server_private_key, server_public_key) for use with the
AKE, along with a Nh-byte oprf_seed. S can use the same pair of keys with multiple
clients and can opt to use multiple seeds (so long as they are kept consistent for
each client). These steps can happen offline, i.e., before the registration phase.

If using `external` mode, the client provides a key pair
(client_private_key, client_public_key)
for an AKE protocol which is suitable for use with OPAQUE; See {{online-phase}}.
The private-public keys (client_private_key, client_public_key) may be randomly
generated (using a cryptographically secure pseudorandom number generator) for the
account or provided by the calling client. Clients MUST NOT use the same key pair
(client_private_key, client_public_key) for two different accounts.

Once complete, the registration process proceeds as follows. The client inputs
the following values:

- password: client password.
- creds: client credentials, as described in {{client-credential-storage}}.

The server inputs the following values:

> - server_private_key: server private for the AKE protocol.

Reviewer 2: ```suggestion
- server_private_key: server private key for the AKE protocol.
```

> - server_public_key: server public for the AKE protocol.

Reviewer 2: ```suggestion
- server_public_key: server public key for the AKE protocol.
```

#### FinalizeRequest {#finalize-request}

To create the user record used for further authentication, the client executes
the following function. In the internal key mode, the `client_private_key` is nil.

Depending on the mode, implementations are free to leave out the `client_private_key`
parameter (`internal` mode), or to additionally include `client_public_key`
(`external` mode). See {#envelope-creation-recovery} for more details.

~~~
FinalizeRequest(client_private_key, password, blind, response)

Input:

> - client_private_key, the client's private key. In the internal key mode, this is nil.

Reviewer 2: ```suggestion
- client_private_key, the client's private key. In internal mode, this is nil.
```

- password, an opaque byte string containing the client's password.
- creds, a Credentials structure.
- blind, the OPRF scalar value used for blinding.
- response, a RegistrationResponse structure.

Output:
- record, a RegistrationUpload structure.
- export_key, an additional client key.

Steps:
1. y = Finalize(password, blind, response.data)
2. random_pwd = Extract("", Harden(y, params))

> 3. envelope, client_public_key, masking_key, export_key =

Reviewer 2: ```suggestion
3. (envelope, client_public_key, masking_key, export_key) =
```

#### Finalize Registration {#finalize-registration}

The server stores the `record` object as the credential file for each client
along with the associated `credential_identifier` and `client_identity` (if
different). Note that the values `oprf_seed` and `server_private_key` from the
server's setup phase must also be persisted.

# Online Authenticated Key Exchange {#online-phase}

The generic outline of OPAQUE with a 3-message AKE protocol includes three messages
ke1, ke2, and ke3, where ke1 and ke2 include key exchange shares, e.g., DH values, sent
by the client and server, respectively, and ke3 provides explicit client authentication and

> full forward security (without it, forward secrecy is only achieved against eavesdroppers

Reviewer: ```suggestion
full forward security (without it, forward secrecy is only achieved against eavesdroppers,
```

- credential_identifier: client credential identifier.
- oprf_seed: seed used to derive per-client OPRF keys.
- server_info: optional, application-specific information to send to the client
  during the handshake.

The client receives two outputs: a session secret and an export key. The export key
is only available to the client, and may be used for additional application-specific
purposes, as outlined in {{export-key-usage}}. The output `export_key` MUST NOT be
used in any way before the protocol completes successfully. See {{envelope-encryption}}
for more details about this requirement. The server receives a single output: a session
secret matching that of the client's.

The protocol runs as shown below:

~~~
  Client                                         Server
 ------------------------------------------------------
  ke1 = ClientInit(client_identity, password, client_info)

                         ke1
              ------------------------->

  ke2 = ServerInit(server_identity, server_private_key,
                    server_public_key, record,
                    credential_identifier, oprf_seed, ke1)

                         ke2
              <-------------------------

>     ke3,
>     server_info,
>     session_key,
>     export_key = ClientFinish(password, client_identity,

Reviewer 2: ```suggestion
    (ke3,
    server_info,
    session_key,
    export_key) = ClientFinish(password, client_identity,
```

#### CreateCredentialResponse {#create-credential-response}

There are two scenarios to handle for the construction of a CredentialResponse
object: either the record for the client exists (corresponding to a properly
registered client), or it was never created (corresponding to a client that has

> yet to register).

Reviewer: ```suggestion
yet to register, or never will).
```

RecoverCredentials(password, blind, response, creds)

Input:
- password, an opaque byte string containing the client's password.
- blind, an OPRF scalar value.
- response, a CredentialResponse structure.
- creds, a Credentials structure.

Output:
- client_private_key, the client's private key for the AKE protocol.
- server_public_key, the public key of the server.
- export_key, an additional client key.

Steps:
1. y = Finalize(password, blind, response.data)
2. random_pwd = Extract("", Harden(y, params))
3. masking_key = Expand(random_pwd, "MaskingKey", Nh)
4. credential_response_pad = Expand(masking_key,
     concat(response.masking_nonce, "CredentialResponsePad"), Npk + Ne)
5. concat(server_public_key, envelope) = xor(credential_response_pad,
                                              response.masked_response)

> 6. client_private_key, export_key =

Reviewer 2: ```suggestion
6. (client_private_key, export_key) =
```

Preamble(client_identity, ke1, server_identity, inner_ke2)

Input:
- client_identity, the optional encoded client identity, which is set
  to client_public_key if not specified.
- ke1, a KE1 message structure.
- server_identity, the optional encoded server identity, which is set
  to server_public_key if not specified.

> - inner_ke2, a inner_ke2 structure as defined in KE2.

Reviewer: ```suggestion
- inner_ke2, an inner_ke2 structure as defined in KE2.
```

- state, a ClientState structure

Input:
- password, an opaque byte string containing the client's password.
- client_identity, the optional encoded client identity, which is set
  to client_public_key if not specified.
- server_identity, the optional encoded server identity, which is set
  to server_public_key if not specified.
- ke1, a KE1 message structure.
- ke2, a KE2 message structure.

Output:
- ke3, a KE3 message structure.
- server_info, optional application-specific information sent encrypted
  and authenticated to the client.
- session_key, the session's shared secret.

Steps:
1. Create Credentials creds with (client_identity, server_identity)

> 2. client_private_key, server_public_key, export_key =

Reviewer 2: ```suggestion
2. (client_private_key, server_public_key, export_key) =
```

    RecoverCredentials(password, state.blind, ke2.CredentialResponse)

> 3. ke3, server_info, session_key =

Reviewer 2: ```suggestion
3. (ke3, server_info, session_key) =
```

### External Server API {#opaque-server}

~~~
ServerInit(server_identity, server_private_key, server_public_key,
           record, credential_identifier, oprf_seed, ke1)

Input:
- server_identity, the optional encoded server identity, which is set to
  server_public_key if nil.
- server_private_key, the server's private key.
- server_public_key, the server's public key.
- server_info, the optional server info sent unencrypted to the client.
- record, the client's RegistrationUpload structure.
- credential_identifier, an identifier that uniquely represents the credential
  being registered.
- oprf_seed, the server-side seed of Nh bytes used to generate an oprf_key.
- ke1, a KE1 message structure.

Output:
- ke2, a KE2 structure.
- client_info, the optional client_info sent unencrypted to the server, only
  authenticated with client_mac in KE3.

Steps:
1. response = CreateCredentialResponse(ke1.request, server_public_key, record,
    credential_identifier, oprf_seed)

> 2. ke2, client_info = Response(server_identity, server_private_key,

Reviewer 2: ```suggestion
2. (ke2, client_info) = Response(server_identity, server_private_key,
```
